### PR TITLE
Store single ScaleUpRequest per node group

### PR DESCRIPTION
**Pull request by losipiuk**

### CODE REVIEW

File: cluster-autoscaler/clusterstate/clusterstate.go [hunk 1]
```diff
@@ -173,48 +183,35 @@ func (csr *ClusterStateRegistry) updateScaleRequests(currentTime time.Time) {
...
-	csr.scaleUpRequests = newSur
-	for _, sur := range timedOutSur {
-		// IsNodeGroupScalingUp returns true if there is another
-		// scale-up still going on for this group, so it's ok for node
-		// group to still have upcoming nodes. If there is no other
-		// scale-up we have VMs that failed to provision within timeout,
-		// so we consider it a failed scale-up
-		if !csr.IsNodeGroupScalingUp(sur.NodeGroupName) {
+
+		if scaleUpRequest.ExpectedAddTime.Before(currentTime) {
```
Comment (aleksandra-malinowska) [inline cluster-autoscaler/clusterstate/clusterstate.go]: I know it's terribly, terribly unlikely, but by using Before() here and After() later, we'll quietly drop requests if ExpectedAddTime == currentTime

Comment (losipiuk) [reply cluster-autoscaler/clusterstate/clusterstate.go]: Good catch! Will fix.

File: cluster-autoscaler/clusterstate/clusterstate.go [hunk 2]
```diff
@@ -173,48 +183,35 @@ func (csr *ClusterStateRegistry) updateScaleRequests(currentTime time.Time) {
...
-		if !csr.areThereUpcomingNodesInNodeGroup(sur.NodeGroupName) {
+	for nodeGroupName, scaleUpRequest := range csr.scaleUpRequests {
+		if !csr.areThereUpcomingNodesInNodeGroup(nodeGroupName) {
 			// scale-out finished successfully
 			// remove it and reset node group backoff
-			csr.nodeGroupBackoffInfo.RemoveBackoff(sur.NodeGroupName)
+			csr.nodeGroupBackoffInfo.RemoveBackoff(nodeGroupName)
 			glog.V(4).Infof("Scale up in group %v finished successfully in %v",
-				sur.NodeGroupName, currentTime.Sub(sur.Time))
+				nodeGroupName, currentTime.Sub(scaleUpRequest.Time))
```
Comment (aleksandra-malinowska) [inline cluster-autoscaler/clusterstate/clusterstate.go]: Do we want to do `delete(csr.scaleUpRequests, nodeGroupName)` here as well, or is it cleaned up elsewhere?

File: cluster-autoscaler/clusterstate/clusterstate.go [hunk 3]
```diff
@@ -406,13 +403,8 @@ func (csr *ClusterStateRegistry) IsNodeGroupScalingUp(nodeGroupName string) bool
...
 	}
-	// Let's check if there is an active scale up request
-	for _, request := range csr.scaleUpRequests {
-		if request.NodeGroupName == nodeGroupName {
-			return true
-		}
-	}
-	return false
+	scaleUpRequest, found := csr.scaleUpRequests[nodeGroupName]
+	return found && scaleUpRequest.Increase > 0
```
Comment (aleksandra-malinowska) [inline cluster-autoscaler/clusterstate/clusterstate.go]: Is there a valid use-case for scaleUpRequest.Increase <= 0?

**Comment by losipiuk** [comment]
Updated

**Pull request merged**